include/seastar/util/closeable.hh

```diff
@@ -61,6 +61,19 @@ public:
     /// Construct an object that will auto-close \c obj when destroyed.
     /// \tparam obj the object to auto-close.
     deferred_close(Object& obj) noexcept : _obj(obj) {}
+    /// Moves the \c deferred_close into a new one, and
+    /// the old one is canceled.
+    deferred_close(deferred_close&& x) noexcept : _obj(x._obj), _closed(std::exchange(x._closed, true)) {}
+    deferred_close(const deferred_close&) = delete;
+    /// Move-assign another \ref deferred_close.
+    /// The current \ref deferred_close is closed before being assigned.
+    /// And the other one's state is transferred to the current one.
+    deferred_close& operator=(deferred_close&& x) noexcept {
+        do_close();
+        _obj = x._obj;
```

Comment:
@tchaikov , this assignment will not re-assign and `Object& _obj` reference, but instead will try to copy referenced objects around, which is not what the `deferred_close` moving is supposed to do

Reply:
ahh, you are right. i just realized that we are not allowed to reassign a reference. see https://isocpp.org/wiki/faq/references#reseating-refs 

as an alternative, we can use `std::reference_wrapper` to hold the reference or just mark the 
```c++
deferred_close& operator=(deferred_close&& x) = delete;
```

@xemul @avikivity what do you think?

Reply:
created #1223 for further discussion.

Reply:
AFAIK, there's a trick

```c++
foo& operator=(foo&& o) {
    if (&o != this) {
        ~foo();
        new (this) foo(std::move(o));
    }
    return *this;
}
```

to implement move-assignment via move-ctor, but I cannot say how ... idiomatically correct it is

Reply:
hmm, i have mixed feelings here: it reuses the move ctor, so yes! but it calls destructor manually and uses placement new, this hurts the readability a little bit, IMHO.

what do you think about `std::reference_wrapper`? it's much simpler.

Reply:
It's a dangerous trick. We can just change _obj to be a pointer.

Reply:
but i don't feel strong either way. please let me know what you prefer and i will update the PR accordingly if necessary.